Use virtualization plugin to detect if we're on openstack on Windows.

jjustice6 opened the pull request:
Signed-off-by: Joshua Justice <jjustice6@bloomberg.net>

<!--- Provide a short summary of your changes in the Title above -->

## Description

If you are using Windows on openstack, the openstack plugin fails to detect that it is using openstack unless you use hints due to the lack of `dmi` on Windows.
It is possible to work around this using `'wmi-lite'` via data contained in `wmi.first_of('Win32_ComputerSystem')`, specifically, the 'manufacturer' and 'model' fields, or data already collected by Ohai elsewhere.

## Related Issue

https://github.com/chef/ohai/issues/1390

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the **CONTRIBUTING** document.
- [X] I have run the pre-merge tests locally and they pass.
- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

jjustice6 commented:
I'm not sure if this is how Chef expects this to be structured, or if I should make an all-new plugin in the windows/ folder for this purpose.

jjustice6 commented on lib/ohai/plugins/openstack.rb:
```diff
@@ -41,6 +42,27 @@ def openstack_dmi?
+  # if we're on windows, use wmi since it doesn't have dmi
+  def openstack_wmi?
+    if os == "windows"
+      require "wmi-lite"
```
Note to self: poking around, I see I should be using `collect_os` here, and my require should be "wmi-lite/wmi"

jjustice6 replied:
Or should it be a separate `collect_data(:windows)`?

tas50 commented:
I think you might want to take a look at depending on the virtualization plugin and then running if that plugin detected the openstack hypervisor. That plugin has pretty similar logic to this already, but generalized for all the different manufacturer IDs

https://github.com/chef/ohai/blob/master/lib/ohai/plugins/windows/virtualization.rb

https://github.com/chef/ohai/blob/master/lib/ohai/mixin/dmi_decode.rb

jjustice6 commented:
Makes sense, will do.

jjustice6 commented:
I've rewritten it to use the info from the virtualization plugin, and added a test accordingly. (A basic one, since there's no need to re-test all the other stuff.)

tas50 commented on lib/ohai/plugins/openstack.rb:
```diff
@@ -61,7 +70,7 @@ def openstack_provider
   collect_data do
     # fetch data if we look like openstack
-    if openstack_hint? || openstack_dmi?
+    if openstack_hint? || openstack_dmi? || openstack_virtualization?
```
I think this DMI check method becomes redundant as long as the virtualization plugin contains all the same manufacturer IDs. I’m on thumbs at the moment so I’ll check on this when I get to a real computer

jjustice6 replied:
I've got no problem making that change in this PR, I can even move the tests to make that work. Just let me know.

jjustice6 replied:
I think it's fine, since it calls the same function under the hood: https://github.com/chef/ohai/blob/master/lib/ohai/plugins/linux/virtualization.rb#L135 - and I've confirmed a linux openstack box does report `"system":"openstack"`.

tas50 replied:
I'll clean it up post merge. There's just no point in the extra logic now that they use the same logic under the hood.

jjustice6 replied:
Sure, thanks.
